Tentative: match new PR about 2E "unless the element is marked as presentational"

giacomo-petri opened the pull request:
Closes: https://github.com/web-platform-tests/interop-accessibility/issues/167

Relates and match: https://github.com/w3c/aria/pull/2405

Ambiguity: https://www.w3.org/TR/accname-1.2/#comp_host_language_label

This PR adds tests for:
- elements receiving the accessible name from host language label elements (both explicit and implicit) with role="presentation";
- elements with role=presentation that do not receive the accessible name from host language label elements (both explicit and implicit) 
- elements with role=presentation that receives the accessible name from host language label elements (both explicit and implicit) due to presentational roles conflict resolution

Note: since I created tests for acc name related to presentational roles conflict resolution, I have also included tests to verify that the computed role aligns with the expected result.

giacomo-petri committed:
Acc test for input with acc name coming from label elements with role presentation

giacomo-petri committed:
Update comp_host_language_label.tentative.html

giacomo-petri committed:
Update comp_host_language_label.tentative.html

giacomo-petri committed:
Create role_none_conflict_resolution.tentative.html

lolaodelola commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,86 @@
+    <h2>HTML input with role="none" and label associated (presentational roles conflict resolution)</h2>
+    <h3>HTML input label/for</h3>
+    <label for="irn">input label</label>
+    <input id="irn" type="text" data-expectedlabel="input label" data-testname="html: label[for]input[role presentation]" class="ex" role="presentation">
```
Hi @giacomo-petri, should the input here have a `role="none"` or are you using `role="presentational"` interchangeably?

cookiecrook replied:
yes, role="none" is preferred, but either should work. role="presentation" or role="none" will both return a computed label of none.

cookiecrook replied:
lola's comment below makes lola's intention here more clear. this is just about editorial comment consistency.

lolaodelola commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,86 @@
+    <h2>HTML input with role="none" and label with role presentation (presentational roles conflict resolution)</h2>
+    <h3>HTML input label/for</h3>
+    <label role="presentation" for="lirp">input label</label>
+    <input id="lirp" type="text" data-expectedlabel="input label" data-testname="html: label[for][role presentation] input[role presentation]" class="ex" role="presentation">
```
Same as above. The heading indicates this is testing for a label with `role="presentation"` and input with `role="none"` but the input here is `role="presentation"`.

lolaodelola commented:
Hi @giacomo-petri, I'm coming from the review of [PR 2405](https://github.com/w3c/aria/pull/2405). This looks good to me for the most part, Scott's addressed some of the initial confusion. I noted some semantic confusion in the naming of the test sections vs what is being tested. While `role="presentation"` and `role="none"` do the same thing, it's confusing to have the test refer to testing `role="none"` on an input but the actual test setup has `role="presentation"` on the input. In most cases where you mention testing for none, you're testing for presentation, I didn't want to litter your inbox with every instance of this but I've noted two above.

Unless there's something I'm missing, the heading and test setup should match to avoid confusion for folks when reading.

Thank you for this work!

giacomo-petri committed:
Update comp_host_language_label.tentative.html

giacomo-petri committed:
Update role_none_conflict_resolution.tentative.html

giacomo-petri commented:
Done

lolaodelola commented:
Thanks @giacomo-petri! You have some lint errors, once you've addressed those I'll be happy to approve this.

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,86 @@
+    </label>
+
+    <h2>HTML img with role="presentation" and non empty alt</h2>
+    <img src="https://www.w3.org/assets/logos/w3c/w3c-no-bars.svg" role="presentation" alt="w3c logo" data-expectedlabel="" data-testname="html: img[role presentation][non-empty alt]">
```
You're missing class="ex" on this one and some or all below. 

Since these files are new, better to change to the new convention: "ex-label" rather than the ambiguous "ex" I started with.

cookiecrook replied:
For future reference, when you forget, you'll notice your results list being a lot shorter than your test list.

giacomo-petri committed:
Update comp_host_language_label.tentative.html

giacomo-petri committed:
Update role_none_conflict_resolution.tentative.html

giacomo-petri commented:
Thanks all:
- replaced .ex with .ex-label (and added where missing)
- replaced all role="presentation" with role="none"

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,86 @@
+    <fieldset data-expectedlabel="legend" data-testname="html: fieldset wrapping legend[role none]" class="ex-label"><legend role="none">legend</legend></fieldset>
+
+    <h2>HTML fieldset with role="none" wrapping legend</h2>
+    <fieldset role="none" data-expectedlabel="" data-testname="html: fieldset[role none] wrapping legend" class="ex-label"><legend>legend</legend></fieldset>
```
This might not be a valid test assumption, as the WebDriver methods don't guarantee a a return value for an element that is ignored by accessibility. That said, maybe it will but we won't know until all the CI bots run. This type of scenario may be more reliable to test once the new WebDriver accessibility extensions land. Ditto for the next one.

cookiecrook replied:
For example, this could be the cause of the Gecko CI crash, though I haven't looked into it yet.

cookiecrook replied:
Or those othe "MISSING" results.

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,86 @@
+    <fieldset role="none" data-expectedlabel="" data-testname="html: fieldset[role none] wrapping legend[role none]" class="ex-label"><legend role="none">legend</legend></fieldset>
+
+    <h2>HTML table wrapping caption with role="none"</h2>
+    <table data-expectedlabel="caption" data-testname="html: table wrapping caption[role none]" class="ex-label"><caption role="none">caption</caption></table>
```
This is an invalid test because it's an invalid table. Simple table tests should contain at least two headers in a header row and two cells in a second row, or you might trigger similar-but-occasionally-different layout table heuristics in different engines. See the table_roles file for comparison. Ditto for the other table tests below.

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,86 @@
+    <table role="none" data-expectedlabel="" data-testname="html: table[role none] wrapping caption" class="ex-label"><caption>caption</caption></table>
+
+    <h2>HTML table role="none" wrapping caption with role="none"</h2>
+    <table role="none" data-expectedlabel="" data-testname="html: table[role none] wrapping caption[role none]" class="ex-label"><caption role="none">caption</caption></table>
```
Like the fieldset[role=none] test above, this may not be a valid test assumption due to an artifact of the WebDriver methods not defining the expected return type when a DOM element no longer has a backing AX element due to the none role. Wait until all the CI bots complete, since there are implementation detail differences this will likely illustrate.

cookiecrook commented on wai-aria/role/role_none_conflict_resolution.tentative.html:
```diff
@@ -0,0 +1,28 @@
+
+    <h2>HTML input with role="none" (presentational roles conflict resolution)</h2>
+    <label for="irn">input label</label>
+    <input id="irn" type="text" data-expectedrole="textbox" data-testname="HTML input[role none] but focusable" class="ex" role="none">
```
```suggestion
    <input id="irn" type="text" data-expectedrole="textbox" data-testname="HTML input[role none] but focusable" class="ex-role" role="none">
```

cookiecrook commented on wai-aria/role/role_none_conflict_resolution.tentative.html:
```diff
@@ -0,0 +1,28 @@
+    <input id="irpd" type="text" data-testname="HTML input[role none][disabled]" class="ex-generic" role="none" disabled>
+
+    <script>
+      AriaUtils.verifyRolesBySelector(".ex");
```
```suggestion
      AriaUtils.verifyRolesBySelector(".ex-role");
```

cookiecrook commented on wai-aria/role/role_none_conflict_resolution.tentative.html:
```diff
@@ -0,0 +1,28 @@
+    <h2>HTML input with role="none" and disabled (presentational roles conflict resolution does not apply due to disabled attr)</h2>
+    <label for="irpd">input label</label>
+    <input id="irpd" type="text" data-testname="HTML input[role none][disabled]" class="ex-generic" role="none" disabled>
+
```
You probably want tests for `input[readonly]` with and w/o `disabled` too, though I don't expect that would have different results from these.

giacomo-petri replied:
Added tests

cookiecrook reviewed (changes requested):
Thanks for the updates! Almost ready!

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,86 @@
+    <img src="https://www.w3.org/assets/logos/w3c/w3c-no-bars.svg" role="none" alt="w3c logo" data-expectedlabel="" data-testname="html: img[role none][non-empty alt]" class="ex-label">
+
+    <h2>SVG wrapping title with role="none"</h2>
+    <svg viewBox="0 0 20 10" xmlns="http://www.w3.org/2000/svg" data-expectedlabel="circle" data-testname="svg: svg wrapping title[role none]" class="ex-label">
+      <circle cx="5" cy="5" r="4">
+        <title role="none">circle</title>
+      </circle>
+    </svg>
```
I'm also not sure about the nameFrom:contents assumption you're making for this test. If the SVG element has its role invalidated with role=none, which spec defined the fallback to a role (like group) that supports nameFrom:contents?

cookiecrook replied:
So this test may be valid if you move the test attributes to the child node:
```suggestion
    <svg viewBox="0 0 20 10" xmlns="http://www.w3.org/2000/svg">
      <circle cx="5" cy="5" r="4" data-expectedlabel="circle" data-testname="svg: svg circle title[role=none]" class="ex-label">
        <title role="none">circle</title>
      </circle>
    </svg>
```

cookiecrook replied:
Neither HTML-AAM or SVG-AAM define a role for the `<svg>` element itself. I'd expect either `group` or `image` and neither of those uses namefrom:contents either.

cookiecrook replied:
Marking this thread as resolved since @giacomo-petri's latest commit includes this.

giacomo-petri committed:
Apply suggestions from code review

Co-authored-by: James Craig <cookiecrook@users.noreply.github.com>

giacomo-petri committed:
Apply suggestions from code review

Co-authored-by: James Craig <cookiecrook@users.noreply.github.com>

giacomo-petri committed:
Update role_none_conflict_resolution.tentative.html

- added readonly testcases with and w/o disabled
- replaced [role none] witg [role=none] in data-testname

giacomo-petri committed:
Update comp_host_language_label.tentative.html

replaced role none with role=none

giacomo-petri committed:
Update comp_host_language_label.tentative.html

added cross reference comment

giacomo-petri committed:
Update role_none_conflict_resolution.tentative.html

add cross reference comment

giacomo-petri committed:
Update comp_host_language_label.tentative.html

expectations moved from svg to circle

giacomo-petri committed:
Update comp_host_language_label.tentative.html

table fix (added THs and TDs)

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,86 @@
+    <h2>HTML input with role="none" and label associated (presentational roles conflict resolution)</h2>
+    <h3>HTML input label/for</h3>
+    <label for="irn">input label</label>
+    <input id="irn" type="text" data-expectedlabel="input label" data-testname="html: label[for]input[role none]" class="ex-label" role="none">
```
Okay to leave as-is. Thanks.

cookiecrook commented:
Fix the [lint errors](https://github.com/web-platform-tests/wpt/pull/50002/checks?check_run_id=35763252359) and the bots will kick off another attempt to get that last CI result.

cookiecrook commented:
For future reference... The link check result is fast, usually just a few minutes, and you always want to correct those errors before continuing with review requests and verifying results.

giacomo-petri committed:
Update comp_host_language_label.tentative.html

giacomo-petri commented:
Lint errors fixed!

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    <h2>HTML fieldset wrapping legend with role="none"</h2>
+    <fieldset data-expectedlabel="legend" data-testname="html: fieldset wrapping legend[role=none]" class="ex-label"><legend role="none">legend</legend></fieldset>
+
+    <h2>HTML fieldset with role="none" wrapping legend</h2>
+    <fieldset role="none" data-expectedlabel="" data-testname="html: fieldset[role=none] wrapping legend" class="ex-label"><legend>legend</legend></fieldset>
```
I don't think this is a valid test, as if the role is none, there is no mandate in WebDriver to have a backing object or return a label. I would remove this test and possible refile as a new "blocked by testability" issue.

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    <h2>HTML fieldset with role="none" wrapping legend</h2>
+    <fieldset role="none" data-expectedlabel="" data-testname="html: fieldset[role=none] wrapping legend" class="ex-label"><legend>legend</legend></fieldset>
+
+    <h2>HTML fieldset role="none" wrapping legend with role="none"</h2>
+    <fieldset role="none" data-expectedlabel="" data-testname="html: fieldset[role=none] wrapping legend[role=none]" class="ex-label"><legend role="none">legend</legend></fieldset>
```
Same

cookiecrook replied:
Note: These are out of order on the Conversation tab... Review comments in order on Review tab.

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    </table>
+
+    <h2>HTML table role="none" wrapping caption with role="none"</h2>
+    <table role="none" data-expectedlabel="" data-testname="html: table[role=none] wrapping caption[role=none]" class="ex-label">
```
Same

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    </table>
+
+    <h2>HTML table with role="none" wrapping caption</h2>
+    <table role="none" data-expectedlabel="" data-testname="html: table[role=none] wrapping caption" class="ex-label">
```
Same as "invalidated" comment above.

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    <h3>HTML label encapsulation</h3>
+    <label>
+      <span>input label</span>
+      <input type="text" data-expectedlabel="input label" data-testname="html: input[role=none] with label encapsulation" class="ex-label" role="none">
```
Same

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    <h2>HTML input with role="none" and label with role none (presentational roles conflict resolution)</h2>
+    <h3>HTML input label/for</h3>
+    <label role="none" for="lirp">input label</label>
+    <input id="lirp" type="text" data-expectedlabel="input label" data-testname="html: label[for][role=none] input[role=none]" class="ex-label" role="none">
```
Same

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    <h3>HTML label encapsulation</h3>
+    <label role="none">
+      <span>input label</span>
+      <input type="text" data-expectedlabel="input label" data-testname="html: input[role=none] with label[role=none] encapsulation" class="ex-label" role="none">
```
Same

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    <h2>HTML input with role="none" and disabled with label associated</h2>
+    <h3>HTML input label/for</h3>
+    <label for="irpd">input label</label>
+    <input id="irpd" type="text" data-expectedlabel="" data-testname="html: label[for]input[role=none][disabled]" class="ex-label" role="none" disabled>
```
Same

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    <h3>HTML label encapsulation</h3>
+    <label>
+      <span>input label</span>
+      <input type="text" data-expectedlabel="" data-testname="html: input[role=none][disabled] with label encapsulation" class="ex-label" role="none" disabled>
```
Same

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    </label>
+
+    <h2>HTML img with role="none" and non empty alt</h2>
+    <img src="https://www.w3.org/assets/logos/w3c/w3c-no-bars.svg" role="none" alt="w3c logo" data-expectedlabel="" data-testname="html: img[role=none][non-empty alt]" class="ex-label">
```
Same

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    <h2>HTML input with role="none" and label associated (presentational roles conflict resolution)</h2>
+    <h3>HTML input label/for</h3>
+    <label for="irn">input label</label>
+    <input id="irn" type="text" data-expectedlabel="input label" data-testname="html: label[for]input[role=none]" class="ex-label" role="none">
```
Is the goal here that the `none` role is invalidated by the input's focusability? If so, say so in the comment or test name. (FYI, test name is not a valid CSS selector. IIRC, you'd need a `+` before `input` to match the adjacent sibling selector.) 

Perhaps `data-testname="html:label[for]+input[role=none] (role invalidated due to focusability)"`

Otherwise, if `none` is the expected role, I don't think this is a valid test, because there is no mandate in WebDriver to have a backing object on an ignored element or to return a label from the non-existent object; not even an empty string. If label from a 'none' role is the intention, I would remove this test and possible refile as a new "blocked by testability" issue.

giacomo-petri replied:
Thanks @cookiecrook!

I'm updating the data-testname with your suggestion since this scenario triggers the presentational roles conflict resolution, and the explicit role set is expected to be ignored due to focusability. Worth mentioning that, in this specific test case, we don't care about the role as it's meant to test only the accessible name. I'm currently trying to address the role topic in the other file: [wai-aria/role/role_none_conflict_resolution.tentative.html](https://github.com/web-platform-tests/wpt/pull/50002/files#diff-701a92ad8707877b8e9fabb34beaa6e85daa5c714bfb58587b40c0c6dabfe0c4).

giacomo-petri replied:
(of course I understand the impact of the role, since the acc name on generic is irrelevant)

giacomo-petri replied:
I've updated all data-testname values with valid CSS selectors for consistency and added explanations for why the label should be computed this way.

cookiecrook replied:
Following up on the second point in the next review here: https://github.com/web-platform-tests/wpt/pull/50002#pullrequestreview-2587331996

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    <fieldset role="none" data-expectedlabel="" data-testname="html: fieldset[role=none] wrapping legend[role=none]" class="ex-label"><legend role="none">legend</legend></fieldset>
+
+    <h2>HTML table wrapping caption with role="none"</h2>
+    <table data-expectedlabel="caption" data-testname="html: table wrapping caption[role=none]" class="ex-label">
+      <caption role="none">caption</caption>
```
@scottaohara Do you concur with this test expectation?

scottaohara replied:
yes. i'd have expected the table to still be named from the caption, even if the caption role was marked as none.

giacomo-petri committed:
updated all data-testname with valid CSS selectors and added expectations/reasons

cookiecrook commented on accname/name/comp_host_language_label.tentative.html:
```diff
@@ -0,0 +1,118 @@
+    </table>
+
+    <h2>HTML table with role="none" wrapping caption</h2>
+    <table role="none" data-expectedlabel="" data-testname="html: table[role=none]>caption (role none set on table affects fieldset acc name)" class="ex-label">
```
This is one of the tests that should be moved to a new "blocked by testability" issue. You can't validly (yet) expect to be able to test computedlabel on an element with no role... No engines compute this... I think Chrome just returns the empty string when there is no backing object, but that's an implementation detail, not a test.

cookiecrook reviewed (changes requested):
The label tests for invalidated role none should stay in this PR. Thanks for adding those comments.

However, any **_label_** test where the same element has an expected role of `none` should be removed from this PR entirely, and added to a new issue (keyword "blocked by testability") in the interop-accessibility repo. We can't commit these into WPT yet, because there is no reliable way to get a label from a non-existent accessibility object (role none/generic/etc). The test would be exposing implementation details, but not failures. 

To my knowledge, no engines compute the label on a non-existent accessibility object... I think Chrome just returns the empty string when there is no backing object, but that's an implementation detail, not a test.

giacomo-petri committed:
removed elements with computed role of none
